### ISSUE TITLE
Slug doesn't work if traits are imported at "random" order

### ISSUE THREAD

**Issue by milewski** [area17/twill#94]
Consider the following:

```php
class DemoRepository extends ModuleRepository {

    use HandleTranslations, HandleSlugs; //<-- here

    public function __construct(Demo $model) {
        $this->model = $model;
    }

}
```
i get the following result:

![image](https://user-images.githubusercontent.com/2874967/49271513-fc565700-f4a8-11e8-9e59-21ec81dc2503.png)

However if i change the order of the traits

```php
class DemoRepository extends ModuleRepository {

    use HandleSlugs, HandleTranslations;  //<-- here

    public function __construct(Demo$model) {
        $this->model = $model;
    }

}
```

I get a different output:

![image](https://user-images.githubusercontent.com/2874967/49271581-45a6a680-f4a9-11e8-9476-f73bf06fa413.png)

This makes me wonder that other parts or even other traits might behave differently when importing them out of a specific order.. one might be overriding the work of the other.... somehow... maybe something to investigate

**Comment by ifox** [comment]
Very good catch @milewski. I assume we did not catch it because we generate every module with our CLI but this is definitely something that we should look into. 

I remember thinking about introducing a priority per trait a while ago, which should prevent those issues where the model state needs to be set in a specific order.

To give you a little bit of background, this is an example of how our traits are called from your own repository:

https://github.com/area17/twill/blob/21e731706d9ef1cf3a9ccd115b1ac279d6c0ae04/src/Repositories/ModuleRepository.php#L380-L387

Introducing a notion of priority in there should be relatively simple.

**Comment by bosix** [comment]
Since I faced the issue as well in #1057 I dig a bit deeper into this.

The function `getFormFieldsHandleSlugs` inside `HandleSlugs.php` add data to the fields array below the key `translations`. The problem is that `$fields['translations']` is unset in `getFormFieldsHandleTranslations` (`HandleTranslations.php`).

From my point of view, it seems easy to fix this specific issue by extending the function `getFormFieldsHandleTranslations` and check for slugs before deleting the whole `translations` key from `$fields`.

```php
public function getFormFieldsHandleTranslations($object, $fields)
    {
        $slugTranslations = $fields['translations']['slug'] ?? null;

        unset($fields['translations']);

        if ($object->translations != null && $object->translatedAttributes != null) {
            foreach ($object->translations as $translation) {
                foreach ($object->translatedAttributes as $attribute) {
                    unset($fields[$attribute]);
                    $fields['translations'][$attribute][$translation->locale] = $translation->{$attribute};
                }
            }
        }

        if ($slugTranslations !== null) {
            $fields = array_merge_recursive(
                ['translations' => ['slug' => $slugTranslations]],
                $fields
            );
        }

        return $fields;
    }
```

The problem is that I am not that familiar with the twill code base so I don't know if there are any other combinations that won't work or even cases when it is not a good idea to just add the slugs to `$fields['translations']` again because of any other side effects.

**Comment by haringsrob** [comment]
This will be solved with #1577
